src/ucs/sys/sock.c

```diff
@@ -151,6 +152,74 @@ ucs_status_t ucs_socket_connect_nb_get_status(int fd)
     return UCS_OK;
 }
 
+static inline ucs_status_t ucs_socket_do_io_nb(int fd, void *data, size_t *length_p,
+                                               ucs_socket_io_func_t io_func, const char *name)
+{
+    ssize_t ret;
+
+    ucs_assert(*length_p > 0);
+
+    ret = io_func(fd, data, *length_p, MSG_NOSIGNAL);
+    if (ucs_unlikely(ret == 0)) {
+        ucs_trace("fd %d is closed", fd);
+        return UCS_ERR_CANCELED; /* Connection closed */
+    } else if (ucs_unlikely(ret < 0)) {
```

Comment:
maybe no need for unlikely here?

Reply:
I would suggest to invert, to handle "likely" branch first then everything else

Reply:
@brminich @evgeny-leksikov fixed